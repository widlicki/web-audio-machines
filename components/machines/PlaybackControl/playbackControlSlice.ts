/* eslint-disable no-param-reassign */
import { createSlice } from '@reduxjs/toolkit'
import type { AppState } from '../../../store/index'

export interface PlaybackState {
  status: 'playing' | 'stopped' | 'paused'
}

const initialState: PlaybackState = {
  status: 'stopped',
}

export const playbackControlSlice = createSlice({
  name: 'playbackControl',
  initialState,
  reducers: {
    pause: (state) => {
      state.status = 'paused'
    },
    play: (state) => {
      state.status = 'playing'
    },
    stop: (state) => {
      state.status = 'stopped'
    },
  },
})

export const { pause, play, stop } = playbackControlSlice.actions

export const selectPlaybackStatus = (state: AppState) =>
  state.playbackControl.status

export default playbackControlSlice.reducer
