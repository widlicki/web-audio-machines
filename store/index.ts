import { configureStore } from '@reduxjs/toolkit'
import playbackControlReducer from '../components/machines/PlaybackControl/playbackControlSlice'
import stepControlReducer from '../components/machines/StepControl/stepControlSlice'

export function makeStore() {
  return configureStore({
    reducer: {
      playbackControl: playbackControlReducer,
      stepControl: stepControlReducer,
    },
  })
}

const store = makeStore()

export default store

export type AppState = ReturnType<typeof store.getState>

export type AppDispatch = typeof store.dispatch
