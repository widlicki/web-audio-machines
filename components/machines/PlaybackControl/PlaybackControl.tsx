import { ToggleButton, ToggleButtonGroup } from '@mui/material'
import { useAppDispatch, useAppSelector } from '../../../hooks/hooks'
import {
  PauseButtonIcon,
  PlayButtonIcon,
  StopButtonIcon,
} from '../../base/Icons/Icons'
import Module from '../../base/Module/Module'
import { pause, play, selectPlaybackStatus, stop } from './playbackControlSlice'

const PlaybackControl = () => {
  const dispatch = useAppDispatch()
  const playbackStatus = useAppSelector(selectPlaybackStatus)

  const handleChange = (e, status) => {
    switch (status) {
      case 'playing':
        dispatch(play())
        break
      case 'paused':
        dispatch(pause())
        break
      case 'stopped':
        dispatch(stop())
        break
      default:
        break
    }
  }

  return (
    <Module title="Playback">
      <ToggleButtonGroup
        value={playbackStatus}
        exclusive
        onChange={handleChange}
      >
        <ToggleButton value="playing" aria-label="Play button">
          <PlayButtonIcon />
        </ToggleButton>
        <ToggleButton value="paused" aria-label="Pause button">
          <PauseButtonIcon />
        </ToggleButton>
        <ToggleButton value="stopped" aria-label="Stop button">
          <StopButtonIcon />
        </ToggleButton>
      </ToggleButtonGroup>
    </Module>
  )
}

export default PlaybackControl
