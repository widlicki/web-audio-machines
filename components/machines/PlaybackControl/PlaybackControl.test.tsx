import { fireEvent, render } from '@testing-library/react'
import { Provider } from 'react-redux'
import { makeStore } from '../../../store'
import PlaybackControl from './PlaybackControl'

let component: any

describe('<PlaybackControl />', () => {
  beforeEach(() => {
    const store = makeStore()
    component = (
      <Provider store={store}>
        <PlaybackControl />
      </Provider>
    )
  })

  it('Renders component according to snapshot', () => {
    const { container } = render(component)
    expect(container).toMatchSnapshot()
  })

  it('Sets playback control to "stopped" by default', () => {
    const { getByRole } = render(component)
    const stopButton = getByRole('button', { name: 'Stop button' })

    fireEvent.click(stopButton)

    expect(stopButton.getAttribute('aria-pressed')).toBe('true')
  })

  it('Sets playback control to "playing" when play button is clicked', () => {
    const { getByRole } = render(component)
    const playButton = getByRole('button', { name: 'Play button' })

    fireEvent.click(playButton)

    expect(playButton.getAttribute('aria-pressed')).toBe('true')
  })

  it('Sets playback control to "paused" when pause button is clicked', () => {
    const { getByRole } = render(component)
    const pauseButton = getByRole('button', { name: 'Pause button' })

    fireEvent.click(pauseButton)

    expect(pauseButton.getAttribute('aria-pressed')).toBe('true')
  })

  it('Sets playback control to "stoped" when stop button is clicked', () => {
    const { getByRole } = render(component)
    const playButton = getByRole('button', { name: 'Play button' })
    const stopButton = getByRole('button', { name: 'Stop button' })

    fireEvent.click(playButton)
    expect(playButton.getAttribute('aria-pressed')).toBe('true')

    fireEvent.click(stopButton)
    expect(stopButton.getAttribute('aria-pressed')).toBe('true')
  })
})
