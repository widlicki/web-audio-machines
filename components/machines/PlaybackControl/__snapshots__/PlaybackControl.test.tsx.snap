// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlaybackControl /> Renders component according to snapshot 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-7mywch-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardHeader-root css-1aseecf-MuiCardHeader-root"
    >
      <div
        class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
      >
        <span
          class="MuiTypography-root MuiTypography-subtitle2 MuiCardHeader-title css-1sr7afs-MuiTypography-root"
        >
          Playback
        </span>
      </div>
    </div>
    <div
      class="MuiCardContent-root css-198pza3-MuiCardContent-root"
    >
      <div
        class="MuiToggleButtonGroup-root css-1gjgmky-MuiToggleButtonGroup-root"
        role="group"
      >
        <button
          aria-label="Play button"
          aria-pressed="false"
          class="MuiButtonBase-root MuiToggleButton-root MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal css-ueukts-MuiButtonBase-root-MuiToggleButton-root"
          tabindex="0"
          type="button"
          value="playing"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="PlayArrowIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M8 5v14l11-7z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          aria-label="Pause button"
          aria-pressed="false"
          class="MuiButtonBase-root MuiToggleButton-root MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal css-ueukts-MuiButtonBase-root-MuiToggleButton-root"
          tabindex="0"
          type="button"
          value="paused"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="PauseIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M6 19h4V5H6v14zm8-14v14h4V5h-4z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          aria-label="Stop button"
          aria-pressed="true"
          class="MuiButtonBase-root MuiToggleButton-root Mui-selected MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal css-ueukts-MuiButtonBase-root-MuiToggleButton-root"
          tabindex="0"
          type="button"
          value="stopped"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="StopIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M6 6h12v12H6z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
